src/wg/format/png/reader.d

```diff
@@ -0,0 +1,926 @@
+    if (result.colorType > PngColorType.max || pngColorTypeToSamples[result.colorType] == 0)
+    {
+        return PngHeaderResult("Invalid color type found");
+    }
+    if (!result.bitDepth.isValid())
+    {
+        return PngHeaderResult("Invalid bit depth found");
+    }
+    import std.bitmanip : swapEndian;
+    result.width = result.width.swapEndian;
+    result.height = result.height.swapEndian;
+
+    return result;
+}
+
+alias ImageResult = Result!ImageBuffer;
+
+/// Loads the given byte array as Png image into an ImageBuffer
+ImageResult loadPng(const ubyte[] file) nothrow
+{
+    PngChunk* chunks;
+    return loadPng(file, getGcAllocator(), null, chunks);
+}
+
+/// Loads the given byte array as Png image into an ImageBuffer using the given allocator
+ImageResult loadPng(const ubyte[] file, Allocator* allocator) nothrow @nogc
+{
+    PngChunk* chunks;
+    return loadPng(file, allocator, null, chunks);
+}
+
+/**
+ * Loads the given byte array as Png image into an ImageBuffer and also loads all other
+ * present png chunks as a linked list of data
+ */
+ImageResult loadPng(const ubyte[] file, out PngChunk* chunks) nothrow
+{
+    return loadPng(file, getGcAllocator(), getGcAllocator(), chunks);
+}
+
+/**
+ * Loads the given byte array as Png image into an ImageBuffer and also loads all other
+ * present png chunks as a linked list of data. It uses the first given allocator for
+ * image data and the other for png chunks. If the other allocator is null it will not
+ * load the chunks.
+ */
+ImageResult loadPng(const ubyte[] file, Allocator* allocator,
+            Allocator* chunkAllocator, out PngChunk* chunks) nothrow @nogc
+{
+    auto pngHeader = loadPngHeader(file);
+    if (pngHeader.error) return ImageResult(pngHeader.error);
+
+    auto loader = PngLoadData(file, pngHeader, allocator, chunkAllocator);
+
+    if (loader.error) return ImageResult(loader.error);
+    
+    while (loader.loadPngChunk()) {}
+
+    if (loader.error) return ImageResult(loader.error);
+
+    return loader.result;
+}
+
+// Structures that holds all the data needed during decoding of PNG and contains operations that does the decoding
+private struct PngLoadData {
+    PngHeaderData info;
+    alias info this;
+
+    const(ubyte)[] data; // raw data
+    PngPaletteEntry[] palette;
+    // We want to have a full size transparency buffer even when actual transparency chunk is shorter
+    // to simplify algorithm for moving it to alpha channel.
+    ubyte[256] transparency;
+    int transLength;
+    ubyte[zlibInflateMemory] zlibBuffer;
+    uint zlibBufferUsed;
+    z_stream zStream;
+    bool unzipDone;
+    Allocator* allocator;
+    Allocator* chunkAllocator;
+    PngChunk* chunks;
+    ubyte[] pixels; // decoded pixels
+    int lineBytes;
+    bool headerAddedToMeta;
+    bool dataFound;
+    uint gamma;
+    Chromaticities chromaticities;
+    bool isSRGB;
+    ImageResult result;
+    string error;
```

Comment:
That's a lot of stuff... does this all need to be retained? Is any of this local to a call-tree? Can any of this just be stack-local data rather than bulking up this struct?

Reply:
I guess I can refactor the code so I can have all zlib stuff as local variables of one method, plus unzipDone and dataFound but for the other stuff I don't see a clear way to do it. Would that be enough?

Reply:
I'm not asking for changes specifically... I'm just curious. I'm slowly reading through the code, but my immediate feeling was that some of this data looks like it might only be used in one place, and perhaps it could be on the stack.

I wonder if we need to allocate unpack buffers with the allocator, or if we can just use stack space? Is there a strong upper bound on the size of the unzip buffer? I recall seeing somewhere it was 44kb? That's a reasonably sized stack buffer.

Reply:
Yes, documentation says its 44kb and currently this structure contains a static zlibBuffer array which is, together with the whole struct placed on stack.